### PR TITLE
expression: fix some problems related to flen

### PR DESCRIPTION
**Pull request by wjhuang2016**
### What problem does this PR solve?

Fix some flen problems.

### What is changed and how it works?

1. Introduce DecimalLength2Precision() and Precision2LengthNoTruncation() to fix flen for the precision of `div`. 
2. Fix wrong calculation for concatWS.
3. Fix flen  for sysDate().

### Check List 

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
 'None'
```

### CODE REVIEW

File: types/helper.go [hunk 1]
```diff
@@ -206,3 +206,25 @@ func strToInt(str string) (int64, error) {
 	}
 	return int64(r), err
 }
+
+// DecimalLength2Precision gets the precision.
+func DecimalLength2Precision(length int, scale int, hashUnsignedFlag bool) int {
+	if scale > 0 {
+		length--
+	}
+	if hashUnsignedFlag || length > 0 {
```
Comment (lzmhhh123) [inline types/helper.go:215]: `hashUnsignedFlag` -> `hasUnsignedFlag`

File: expression/builtin_cast.go [hunk 2]
```diff
@@ -1896,6 +1896,9 @@ func WrapWithCastAsDecimal(ctx sessionctx.Context, expr Expression) Expression {
 	}
 	tp := types.NewFieldType(mysql.TypeNewDecimal)
 	tp.Flen, tp.Decimal = expr.GetType().Flen, expr.GetType().Decimal
+	if tp.Flen < 10 {
+		tp.Flen = 10
+	}
```
Comment (wshwsh12) [inline expression/builtin_cast.go:1901]: I can't understand why we need add the check... Could your give me an example?

File: types/field_type.go [hunk 3]
```diff
@@ -293,6 +293,8 @@ func DefaultTypeForValue(value interface{}, tp *FieldType, char string, collate
 		tp.Tp = mysql.TypeNewDecimal
 		tp.Flen = len(x.ToString())
 		tp.Decimal = int(x.digitsFrac)
+		// Add the length for `.`.
+		tp.Flen++
```
Comment (wshwsh12) [inline types/field_type.go:297]: x.ToString() have contained the ".", I think..

**Review by wshwsh12 — COMMENTED** [review]
Rest LGTM

**Review by wshwsh12 — APPROVED** [review]

**Pull request merged**